src/Utility/ResponseTools.cs

```diff
@@ -0,0 +1,313 @@
+    public async Task AddMcpToolsAsync(McpClient client)
+    {
+        if (client == null) throw new ArgumentNullException(nameof(client));
+        _mcpClientsByEndpoint[client.Endpoint.AbsoluteUri] = client;
+        await client.StartAsync().ConfigureAwait(false);
+        BinaryData tools = await client.ListToolsAsync().ConfigureAwait(false);
+        await AddToolsAsync(tools, client).ConfigureAwait(false);
+        _mcpClients.Add(client);
+    }
+
+    /// <summary>
+    /// Adds a remote MCP server as a tool provider.
+    /// </summary>
+    /// <param name="mcpEndpoint">The URI endpoint of the MCP server.</param>
+    /// <returns>A task representing the asynchronous operation.</returns>
+    public async Task AddMcpToolsAsync(Uri mcpEndpoint)
+    {
+        var client = new McpClient(mcpEndpoint);
+        await AddMcpToolsAsync(client).ConfigureAwait(false);
+    }
+
+    private async Task AddToolsAsync(BinaryData toolDefinitions, McpClient client)
+    {
+        using var document = JsonDocument.Parse(toolDefinitions);
+        if (!document.RootElement.TryGetProperty("tools", out JsonElement toolsElement))
+            throw new JsonException("The JSON document must contain a 'tools' array.");
+
+        var serverKey = client.Endpoint.Host + client.Endpoint.Port.ToString();
+        List<ResponseTool> toolsToVectorize = new();
+
+        foreach (var tool in toolsElement.EnumerateArray())
+        {
+            var name = $"{serverKey}{ToolsUtility.McpToolSeparator}{tool.GetProperty("name").GetString()!}";
+            var description = tool.GetProperty("description").GetString()!;
+#pragma warning disable IL2026, IL3050
+            var inputSchema = JsonSerializer.Serialize(
+                JsonSerializer.Deserialize<JsonElement>(tool.GetProperty("inputSchema").GetRawText()),
+                new JsonSerializerOptions { Encoder = System.Text.Encodings.Web.JavaScriptEncoder.UnsafeRelaxedJsonEscaping });
+#pragma warning restore IL2026, IL3050
+
+            var responseTool = ResponseTool.CreateFunctionTool(name, description, BinaryData.FromString(inputSchema));
+            _tools.Add(responseTool);
+            toolsToVectorize.Add(responseTool);
+            _mcpMethods[name] = client.CallToolAsync;
+        }
+
+        if (_client != null)
+        {
+            var embeddings = await _client.GenerateEmbeddingsAsync(toolsToVectorize.ConvertAll(GetDescription)).ConfigureAwait(false);
+            foreach (var embedding in embeddings.Value)
+            {
+                var vector = embedding.ToFloats();
+                var item = toolsToVectorize[embedding.Index];
+                var toolDefinition = SerializeTool(item);
+                _entries.Add(new VectorbaseEntry(vector, toolDefinition));
+            }
+        }
+    }
+
+    private string GetDescription(ResponseTool tool) => (tool as InternalResponsesFunctionTool)?.Description ?? "";
+
+    private BinaryData SerializeTool(ResponseTool tool)

```

Comment:
even this one could be shared if it took 3 parameters (name, description, parameters)